# U.S. Medical Insurance Costs

I decided to use the hint structure to do this. I ran the code for this once or twice before now, but I want to use instructions for this try.

## Goals:
* Find the average age of patients in dataset
* Where are the majority of the individuals from
* Costs of smokers vs non smokers
* What is the average age for someone who has at least one child

*This project will become my side project as time goes on. I will come back and add to it as I learn new things or want to practice new ideas.*

## Importing things I need

*I am only importing things I have learned about so far. Sorry, no panda or pymath*

In [104]:
import pprint
import csv

## Saving Data to variables

Here, I don't know if I want it in dictionaries or just in lists. Lists seem easier to work with, but on the other hand, dictionaries are beefy and I like to use them.

I decided on keeping them as lists that I can combine and seperate as needed.

In [105]:
ages = []
sexes = []
bmis = []
children = []
smokers = []
regions = []
charges = []

def import_data(lst, csv_file, column):
    with open(csv_file) as csv_file:
        reader = csv.DictReader(csv_file)
        for row in reader:
            lst.append(row[column])

import_data(ages, 'insurance.csv', 'age')
import_data(sexes, 'insurance.csv', 'sex')
import_data(bmis, 'insurance.csv', 'bmi')
import_data(children, 'insurance.csv', 'children')
import_data(smokers, 'insurance.csv', 'smoker')
import_data(regions, 'insurance.csv', 'region')
import_data(charges, 'insurance.csv', 'charges')

## Make a class
I want to make a patient class that can hold the things I need it to do. So that is what I am gonna do.

### Defining the calss.
* Giving it the initial data it needs

### Getting to work on the goals = avarage age of patients in dataset
* Get all the ages together and devide them by lenght of ages
* *In the future I want to add the min and max ranges. Also, I am learning more math skills and stat skills from khan academy and such. Will implement the smart folk stuff later.*

### Where are the majority of the people from:
* Make a dictionary with the locations as the keyes and a value that counts the places

### Smokers vs Non-smokers:
* Get the average cost of both
    * Get the cost and the status linked together
* Return with x pays more than y
* *Future x pays z more than y.*

In [106]:
class PatientData:
    def __init__(self, patient_age, patient_sex, patient_bmi, patient_children, patient_smokers, patient_regions, patient_charges):
        self.age = [int(i) for i in patient_age]
        self.sex = patient_sex
        self.bmi = patient_bmi
        self.children = [int(i) for i in patient_children]
        self.smokers = patient_smokers
        self.regions = patient_regions
        self.charges = patient_charges
    
    def analyze_ages(self):
        total_age = 0
        for age in self.age:
            total_age += age
        return "The average age in this dataset is " + str(round(total_age / len(self.age),2)) + " years."
    
    def analyze_regions(self):
        dict_regions = {} #I wanted to do something with this, but ran out of time. Will come back to print info on all places later.
        max_pop = 0
        region = ''
        for place in self.regions:
            dict_regions[place] = self.regions.count(place)
        for place, value in dict_regions.items():
            if value > max_pop:
                max_pop = value
                region = place
        return ("The most people come from the " + region.title() + ".")
    
    def analyze_smokers(self):
        smokers = {'yes' : [],
                  'no' : []}
        lst_smokers = list(zip(self.smokers, self.charges))
        for i in lst_smokers:
            if 'yes' in i:
                smokers['yes'].append(float(i[1]))
            else:
                smokers['no'].append(float(i[1]))
        smoker_average = round(float(sum(smokers['yes']) / len(smokers['yes'])),2)
        non_smoker_average = round(float(sum(smokers['no']) / len(smokers['no'])),2)
        return "Smokers pay an average of: " + str(smoker_average) + " dollars, where as non-smokers pay an average of " + str(non_smoker_average) + " dollars."
        
    def analyze_children(self):
        num_children = {0 : [], 1:[], 2:[], 3:[], 4:[],5:[], 'more':[]}
        lst_children = list(zip(self.children, self.age))
        for child in lst_children:
            if child[0] == 0:
                num_children[0].append(child[1])
            elif child[0] == 1:
                num_children[1].append(child[1])
            elif child[0] == 2:
                num_children[2].append(child[1])
            elif child[0] == 3:
                num_children[3].append(child[1])
            elif child[0] == 4:
                num_children[4].append(child[1])
            elif child[0] == 5:
                num_children[5].append(child[1])
            else:
                num_children['more'].append(child[1])
        return "The average age for a person with 1 child is: " + str(round(float(sum(num_children[1]) / len(num_children[1])),2)) + " years old"
        
    

patient_data = PatientData(ages, sexes, bmis, children, smokers, regions, charges)

## Calling the class

In [107]:
print(patient_data.analyze_ages())
print()
print(patient_data.analyze_regions())
print()
print(patient_data.analyze_smokers())
print()
print(patient_data.analyze_children())

The average age in this dataset is 39.21 years.

The most people come from the Southeast.

Smokers pay an average of: 32050.23 dollars, where as non-smokers pay an average of 8434.27 dollars.

The average age for a person with 1 child is: 39.45 years old


## Stating the findings:
* The avergae age across the data is 39.21 years.
* Most of the patients come from the Southeast
* Smokers pay more than non-smokers
* The average age for a person with 1 child is 39.45 years old

# Plans for future things
This code might not be pretty, or perfect. But it is the first big project I have done in DS ever and I am proud of it. Every cell in my body is screaming to make it perfect, but that will take time away from learning the actual things that will make this project better.

If you are reading this, thank you for taking your time. If you have advice for me, please let me know. I will gladly learn from you.

### Strugles:
I struggled with the last two things a lot. The smoker and the num_children things. I knew what I wanted to do, but just could not make it happen. I feel like I should have done it in fewer lines. It feels like a Frankenstein monster of sorts, but I did not want to get bogged down for ever. (Not gonna give an exact time, but it took me way to long to do them.)

The regions thing also does not feel optimal. It is okay, but it should be better. I am going to see how other people handled it without importing extra things.